core-services/src/main/java/org/zalando/nakadi/partitioning/UserDefinedPartitionStrategy.java

```diff
@@ -10,20 +11,19 @@
 public class UserDefinedPartitionStrategy implements PartitionStrategy {
 
     @Override
-    public String calculatePartition(final EventType eventType, final JSONObject event, final List<String> partitions)
+    public String calculatePartition(final PartitionData partitionData, final List<String> partitions)
             throws PartitioningException {
-        try {
-            final String partition = event.getJSONObject("metadata").getString("partition");
-            if (partitions.contains(partition)) {
-                return partition;
-            } else {
-                throw new PartitioningException("Failed to resolve partition. " +
-                        "Invalid partition specified when publishing event.");
-            }
+        if (Strings.isNullOrEmpty(partitionData.getPartition())) {
+            throw new PartitioningException("Failed to resolve partition. " +
+                    "Failed to get partition from event metadata");
```

Comment:
isn't it to late here to say about metadata? the method works with PartitionData.

Reply:
PartitionData can be missing the partition number since it is only a representation of partition data that comes with the request. 

For instance for publishing requests of event types with a random partition strategy, we don't require the partition numbers to be on the request metadata, and the partition number in metadata is only requested for user-defined partition strategy.

Reply:
After all, I implemented the logic of extracting the partitioning data into all partition strategies https://github.com/zalando/nakadi/pull/1397/files#diff-e2daf1f8029ae29b6cff742e0ac068578328bef4bebc7e6e28a715a1631d15adR31